File: android/lib/services/order_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../app_config.dart';
import '../dtos/order_dto.dart';
import '../responses/order_response.dart';

class OrderService{
  Future<void> placeOrder(OrderDto orderDto) async {
    try {
      final url = Uri.parse('${AppConfig.baseUrl}/orders');
      final response = await http.post(
        url,
        headers: {
          'Content-Type': 'application/json; charset=UTF-8',
          'Accept': 'application/json; charset=UTF-8',
        },
        body: jsonEncode(orderDto.toJson()),
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to place order: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to place order: $e');
    }
  }

  Future<List<OrderResponse>> getOrderById(int orderId) async {
    try {
      final url = Uri.parse('${AppConfig.baseUrl}/orders/$orderId');
      final response = await http.get(
        url,
        headers: {
          'Content-Type': 'application/json; charset=UTF-8',
          'Accept': 'application/json; charset=UTF-8',
        },
      );

      if (response.statusCode == 200) {
        final Map<String, dynamic> json =jsonDecode(response.body);
        final List<dynamic> data = json['data'];
        return data.map((json)=>OrderResponse.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load order detail: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error fetching order detail: $e');
    }
  }

  Future<void> updateOrderStatus(String vnp_TxnRef,String status) async {
    try {
      final url = Uri.parse('${AppConfig.baseUrl}/orders/status').replace(
        queryParameters: {
          'status': status.toString(),
          'vnpTxnRef': vnp_TxnRef.toString(),
        },
      );
      final response = await http.put(
        url,
        headers: {
          'Content-Type': 'application/json; charset=UTF-8',
          'Accept': 'application/json; charset=UTF-8',
        },
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to update status order: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Failed to update status order: $e');
    }
  }

  Future<List<OrderResponse>> getOrdersByUserId(int userId) async {
    try {
      final url = Uri.parse('${AppConfig.baseUrl}/orders/user/$userId');
      final response = await http.get(
        url,
        headers: {
          'Content-Type': 'application/json; charset=UTF-8',
          'Accept': 'application/json; charset=UTF-8',
        },
      );

      if (response.statusCode == 200) {
        final Map<String, dynamic> json =jsonDecode(response.body);
        final List<dynamic> data = json['data'];
        return data.map((json)=>OrderResponse.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load order detail: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error fetching order detail: $e');
    }
  }
}

File: frontend/src/app/services/order.service.ts
import { Injectable } from "@angular/core";
import { environment } from "../environments/environment";
import { HttpClient, HttpParams } from "@angular/common/http";
import { Observable } from "rxjs";
import { ApiResponse } from "../responses/api.response";
import { OrderDto } from "../dtos/order.dto";

@Injectable({
    providedIn: 'root'
})

export class OrderService {
    private apiCreateOrder = `${environment.apiBaseUrl}/orders`;
    constructor(private http: HttpClient) { }

    placeOrder(OrderDto: OrderDto): Observable<ApiResponse> {
        return this.http.post<ApiResponse>(this.apiCreateOrder, OrderDto);
    }

    getOrderById(orderId: number): Observable<ApiResponse> {
        return this.http.get<ApiResponse>(`${environment.apiBaseUrl}/orders/${orderId}`);
    }

    updateOrderStatus(vnp_TxnRef: string, status: string): Observable<ApiResponse> {
        debugger
        const params = new HttpParams()
            .set('status', status.toString())
            .set('vnpTxnRef', vnp_TxnRef.toString());
        const url = `${environment.apiBaseUrl}/orders/status`;
        return this.http.put<ApiResponse>(url, null, { params });
    }

    viewFile(fileName: string) {
        return `${this.apiCreateOrder}/files/${fileName}`;
    }

    getOrdersByUserId(userId: number): Observable<ApiResponse> {
        return this.http.get<ApiResponse>(`${this.apiCreateOrder}/user/${userId}`);
    }

    cancelOrder(id: number): Observable<ApiResponse> {
        debugger
        const url = `${this.apiCreateOrder}/cancel/${id}`;
        return this.http.put<ApiResponse>(url, null);
    }
}